packages/base/src/commonMain/kotlin/elide/struct/RedBlackTreeMap.kt

```diff
@@ -0,0 +1,498 @@
+
+    if (sibling.isRed) {
+      // red sibling: recolor parent and sibling, rotate parent, then retry
+      parent.color = RED
+      sibling.color = BLACK
+      rotate(parent, !sibling.direction)
+
+      rebalanceRemoved(node)
+      return
+    }
+
+    if (sibling.left.isBlack && sibling.right.isBlack) {
+      // black sibling with all-black children, recolor and recurse
+      sibling.color = RED
+
+      if (parent.isRed) parent.color = BLACK
+      else rebalanceRemoved(parent)
+
+      return
+    }
+
+    // sibling has at least one red child, find it
+    val siblingDirection = sibling.direction
+    val redDirection = (sibling.right.isRed)
+
+    if (siblingDirection == redDirection) {
+      // left-left/right-right
+      sibling.color = parent.color
+      sibling.child(redDirection)!!.color = sibling.color
+      rotate(parent, !siblingDirection)
+    } else {
+      // right-left/left-right
+      sibling.child(redDirection)!!.color = parent.color
+      rotate(sibling, siblingDirection)
+      rotate(parent, !siblingDirection)
+    }
+
+    parent.color = BLACK
+  }
+
+  /**
+   * Rotate a [node] in the given [direction] (to the right if `true`, to the left if `false`). The [node] _must_ have
+   * a child in the direction opposing the rotation (a left child for right-rotation, right child for left-rotation),
+   * which is used as the "pivot" for the operation, otherwise an exception will be thrown.
+   *
+   * Rotation affects the node's parent and its "pivot" child, the following figure visually depicts left-rotation
+   * (right-rotation is a mirror case and has the same result, only in the opposite direction):
+   *
+   * ```
+   *  initial  | reparent N, R | reparent <CR> | reparent [N] (final)
+   * ----------|---------------|---------------|---------------------
+   *     /     |            /  |            /  |         /
+   *   [N]     |   [N]    (R)  |   [N]    (R)  |       (R)
+   *   / \     |   /      /    |   / \         |       /
+   *  L  (R)   |  L     <CR>   |  L <CR>       |     [N]
+   *     /     |               |               |     / \
+   *   <CR>    |               |               |    L <CR>
+   * ----------|------(1)------|------(2)------|-----(3)-------------
+   * ↑ rotate [N] to the left
+   * ```
+   *
+   * > See this method's source code for implementation details on (1), (2), and (3).
+   *
+   * Rotation is used to maintain the properties of the Red/Black Tree after a node is [inserted][rebalanceAdded] or
+   * [removed][rebalanceRemoved], and can be paired with recoloring when necessary.
+   */
+  private fun rotate(node: Node<K, V>, direction: Boolean) {
+    val pivot = node.child(!direction) ?: error("node is missing a ${if (direction) "left" else "right"} child")
+    val orphan = pivot.child(direction)
+
+    node.parent?.replace(node, pivot) // (1)
+    if (node == root) root = pivot
+
+    node.setChild(!direction, orphan) // (2)
+    pivot.setChild(direction, node) // (3)
+  }
+}
```

Comment:
New rotation method with visual documentation :D

Reply:
This is amazing. For once I understand Red Black trees lol